File: src/readin.py
# -*- coding: utf-8 -*-
"""
Created on Mon Feb 22 00:58:01 2022

@author: haoxiang
"""

'''
input the data of production planning problem
'''

import pandas as pd
import numpy as np
import copy


# We need data structure for items, plants, calendar

def readin_csv(file_add):
    return pd.read_csv(file_add)


# %%
# item related data
def input_item(item_file):
    # input dm_df_item: item basic information
    # output the item list and holding costs
    item_data = readin_csv(item_file)
    item_list = list(item_data.item_code)  # list of items
    holding_cost = {item_list[i]: item_data.holding_cost[i] for i in range(len(item_list))}
    penalty_cost = {item_list[i]: item_data.penalty_cost[i] for i in range(len(item_list))}
    return item_list, holding_cost, penalty_cost

def input_set(set_file):
    # input dm_df_set: set-capacity_type information
    set_data = readin_csv(set_file)
    item_list = list(set_data.item_code)
    set_dict = {(set_data.item_code[i], set_data.plant[i], set_data.capacity_type[i]): set_data.set[i] for i in
                range(len(item_list))}
    return set_dict


def input_item_set(item_set_file):
    # input dm_df_item_set: item set information
    # output the item production set
    item_set_data = readin_csv(item_set_file)
    item_list = list(item_set_data.item_code)  # list of items
    set_list = list(item_set_data.set.unique())
    item_set = {}
    for i in range(len(item_list)):
        if not (item_list[i] in item_set.keys()):
            item_set[item_list[i]] = [item_set_data.set[i]]
        else:
            item_set[item_list[i]].append(item_set_data.set[i])
    return item_set, set_list


def input_item_alternate(alternate_file):
    # input dm_df_alternate_item: item replacement information
    # output the alternate item production multipliers
    alter_data = readin_csv(alternate_file)
    subbed_list = list(alter_data.item_code1)  # list of items being subbed
    sub_list = list(alter_data.item_code2)  # list of items to sub
    plant_list = list(alter_data.plant)  # list of plants
    period_list = list(alter_data.period)  # list of periods
    priority_list = list(alter_data.priority_col)  # list of directions
    alt_list = []
    alt_dict = {}
    alt_cost = {}
    for i in range(len(subbed_list)):
        if priority_list[i] == 2:
            alt_list.append((plant_list[i], period_list[i], (subbed_list[i], sub_list[i])))
            alt_dict[(plant_list[i], period_list[i], (subbed_list[i], sub_list[i]))] = alter_data.multiplier[i]
            alt_cost[(plant_list[i], period_list[i], (subbed_list[i], sub_list[i]))] = alter_data.alter_cost[i]
        elif priority_list[i] == 1:
            alt_list.append((plant_list[i], period_list[i], (sub_list[i], subbed_list[i])))
            alt_dict[(plant_list[i], period_list[i], (sub_list[i], subbed_list[i]))] = alter_data.multiplier[i]
            alt_cost[(plant_list[i], period_list[i], (sub_list[i], subbed_list[i]))] = alter_data.alter_cost[i]
    return alt_list, alt_dict, alt_cost


def input_unit_capacity(unitC_file):
    # input dm_df_unit_capacity: item unit capacity consumption information
    # output unit capacity, unit capacity type
    unit_cap_data = readin_csv(unitC_file)
    item_list = list(unit_cap_data.item_code)  # list of items
    unit_cap_set = {(item_list[i], unit_cap_data.capacity_type[i]): unit_cap_data.unit_capacity[i] for i in
                    range(len(item_list))}
    return unit_cap_set


def input_item_plant(item_plant_file):
    # input dm_df_item_plant: item plant information
    # output item plant mapping
    item_plant_data = readin_csv(item_plant_file)
    item_list = list(item_plant_data.item_code)
    item_set = set(item_plant_data.item_code)
    item_plant = {item: [] for item in item_set}
    for i in range(len(item_list)):
        if not (item_plant_data.plant[i] in item_plant[item_plant_data.item_code[i]]):
            item_plant[item_plant_data.item_code[i]].append(item_plant_data.plant[i])
    return item_plant


def input_bom(bom_file):
    # input dm_df_bom: item bom relationship information
    # output item bom relationship
    bom_data = readin_csv(bom_file)
    bom_key = {}
    bom_dict = {}
    for i in range(len(bom_data.assembly)):
        if bom_data.plant[i] in bom_key.keys():
            # build up the data structure at the plant level
            bom_key[bom_data.plant[i]].append((bom_data.assembly[i], bom_data.component[i]))
            bom_dict[bom_data.plant[i]][(bom_data.assembly[i], bom_data.component[i])] = bom_data.qty[i]
        else:
            bom_key[bom_data.plant[i]] = [(bom_data.assembly[i], bom_data.component[i])]
            bom_dict[bom_data.plant[i]] = {}
            bom_dict[bom_data.plant[i]][(bom_data.assembly[i], bom_data.component[i])] = bom_data.qty[i]
    return bom_key, bom_dict


def input_production(production_file):
    # input dm_df_production: item production relationship information
    production_data = readin_csv(production_file)
    prod_key = []
    lot_size = {}
    lead_time = {}
    component_holding_cost = {}
    prod_cost = {}
    min_prod = {}
    max_prod = {}
    for i in range(len(production_data.item_code)):
        line_key = (production_data.item_code[i], production_data.plant[i])
        prod_key.append(line_key)
        lot_size[line_key] = production_data.lot_size[i]
        lead_time[line_key] = production_data.lead_time[i]
        component_holding_cost[line_key] = production_data.component_holding_cost[i]
        prod_cost[line_key] = production_data.production_cost[i]
        min_prod[line_key] = production_data.min_production[i]
        max_prod[line_key] = production_data.max_production[i]
    return prod_key, lot_size, lead_time, component_holding_cost, prod_cost, min_prod, max_prod


# %%
# plant related data
def input_plant(plant_file):
    # input dm_df_plant: plant basic information
    plant_data = readin_csv(plant_file)
    plant_list = list(plant_data.plant)
    return plant_list


def input_capacity(cap_file):
    # input dm_df_max_capacity: plant capacity information
    cap_data = readin_csv(cap_file)
    period_list = list(cap_data.period.unique())
    cap_key = []
    max_cap = {}
    for i in range(len(cap_data.set)):
        line_key = (cap_data.set[i], cap_data.plant[i])
        if line_key not in cap_key:
            cap_key.append(line_key)
        if not (line_key in max_cap.keys()):
            max_cap[line_key] = {cap_data.period[i]: cap_data.max_capacity[i]}
        else:
            max_cap[line_key][cap_data.period[i]] = cap_data.max_capacity[i]
    return period_list, cap_key, max_cap


def input_transit(transit_file):
    # input dm_df_transit: plant transit information
    transit_data = readin_csv(transit_file)
    transit_data.drop_duplicates(inplace=True, ignore_index=True)
    transit_data.fillna(0)
    item_transit_list = [
        (transit_data.item_code[i], transit_data.src_plant[i], transit_data.dest_plant[i]) for i in
        range(len(transit_data.item_code))]
    transit_time = {
        (transit_data.item_code[i], transit_data.src_plant[i], transit_data.dest_plant[i]): transit_data.lead_time[i]
        for i in range(len(transit_data.item_code))}
    transit_cost = {
        (transit_data.item_code[i], transit_data.src_plant[i], transit_data.dest_plant[i]): transit_data.transit_cost[i]
        for i in range(len(transit_data.item_code))}
    return item_transit_list, transit_time, transit_cost


def input_init_inv(init_inv_file):
    # input dm_df_inv: plant initial inventory information
    init_inv_data = readin_csv(init_inv_file)
    init_inv = {(init_inv_data.item_code[i], init_inv_data.plant[i]): init_inv_data.qty[i] for i in
                range(len(init_inv_data.item_code))}
    return init_inv


# %%
# time related data
def input_periods(periods_file):
    # input dm_df_periods: periods information
    periods_data = readin_csv(periods_file)
    return list(periods_data.period)


def input_po(po_file):
    # input dm_df_po: external purchase information
    po_data = readin_csv(po_file)
    external_purchase = {(po_data.item_code[i], po_data.plant[i], po_data.period[i]): po_data.qty[i] for i in
                         range(len(po_data.period))}
    return external_purchase


def input_demand(demand_file):
    # input dm_df_demand: external demand information
    demand_data = readin_csv(demand_file)
    real_demand = {(demand_data.item_code[i], demand_data.period[i]): demand_data.order_demand[i] for i in
                   range(len(demand_data.item_code))}
    #forecast_demand = {(demand_data.item_code[i], demand_data.period[i]): demand_data.forecast_demand_with_fr[i] for i
    #                   in range(len(demand_data.item_code))}
    #return real_demand, forecast_demand
    return real_demand


def timeline_adjustment(gbv):
    # adjust the gbv's data time to start from 1 with a difference of 1
    time_start = min(gbv.period_list)
    time_end = max(gbv.period_list)
    time_sorted = sorted(gbv.period_list)

    # set up the new list of periods and construct a dictionary for period lookup
    period_list = [t for t in range(1, time_end - time_start + 2)]
    period_dict = {}
    for t in time_sorted:
        period_dict[t] = t - time_start + 1
    gbv.period_list = period_list
    gbv.period_dict = period_dict
    gbv.T = len(gbv.period_list)

    # change the data in gbv which contains the time aspect
    # cap_period_list
    cap_period_list = [gbv.period_dict[t] for t in gbv.cap_period_list]
    gbv.cap_period_list = cap_period_list
    # maximum capacity
    max_cap = {}
    for ckey in gbv.max_cap.keys():
        max_cap[ckey] = {}
        for t in gbv.max_cap[ckey].keys():
            max_cap[ckey][gbv.period_dict[t]] = gbv.max_cap[ckey][t]
    gbv.max_cap = max_cap

    # alternative list
    alt_list = []
    for item in gbv.alt_list:
        if not ((item[0], gbv.period_dict[item[1]], item[2]) in alt_list):
            alt_list.append((item[0], gbv.period_dict[item[1]], item[2]))
    alt_dict = {}
    alt_cost = {}
    for akey in gbv.alt_dict.keys():
        akey_update = (akey[0], gbv.period_dict[akey[1]], akey[2])
        alt_dict[akey_update] = gbv.alt_dict[akey]
        alt_cost[akey_update] = gbv.alt_cost[akey]
    gbv.alt_list = alt_list
    gbv.alt_dict = alt_dict
    gbv.alt_cost = alt_cost

    # demand
    real_demand = {}
    forecast_demand = {}
    for dkey in gbv.real_demand.keys():
        dkey_update = (dkey[0], gbv.period_dict[dkey[1]])
        real_demand[dkey_update] = gbv.real_demand[dkey]
        #forecast_demand[dkey_update] = gbv.forecast_demand[dkey]
    gbv.real_demand = real_demand
    #gbv.forecast_demand = forecast_demand

    # external purchase
    external_purchase = {}
    for pkey in gbv.external_purchase:
        pkey_update = (pkey[0], pkey[1], gbv.period_dict[pkey[2]])
        external_purchase[pkey_update] = gbv.external_purchase[pkey]
    gbv.external_purchase = external_purchase

    return gbv


def cap_adjustment(gbv):
    unit_cap = {}
    item_plant_set_list = {}
    for skey in gbv.set_dict.keys():
        unit_cap[skey[0], skey[1], gbv.set_dict[skey]] = gbv.unit_cap_set[skey[0], skey[2]]
        if (skey[0], skey[1]) in item_plant_set_list.keys():
            item_plant_set_list[skey[0], skey[1]].append(gbv.set_dict[skey])
        else:
            item_plant_set_list[skey[0], skey[1]] = [gbv.set_dict[skey]]

    gbv.unit_cap = unit_cap
    gbv.item_plant_set_list = item_plant_set_list
    return gbv


def obtain_bounds(gbv):
    # obtain the upper bound for an item at all locations without production, inventory upper bound i, t
    X_list = {}
    for i in gbv.item_list:
        X_list[i] = sum([gbv.init_inv[i, j] for j in gbv.plant_list if (i, j) in gbv.init_inv.keys()])
    X_po_list = {}
    for i in gbv.item_list:
        X_po_list[i] = np.ones(len(gbv.period_list) + 1) * X_list[i]
        for t in range(len(gbv.period_list) + 1):
            X_po_list[i][t:] += sum(
                [gbv.external_purchase[i, j, t] for j in gbv.plant_list if (i, j, t) in gbv.external_purchase.keys()])

    # obtain the upper bound for an item at any specific location without production, inventory upper bound i, j, t
    X_j_po_list = {}
    for i in gbv.item_list:
        for j in gbv.plant_list:
            X_j_po_list[i, j] = np.ones(len(gbv.period_list) + 1) * gbv.init_inv.get((i, j), 0)
            for t in range(len(gbv.period_list) + 1):
                X_j_po_list[i, j][t:] += gbv.external_purchase.get((i, j, t), 0)
    X_bar_j = {}
    for i in gbv.item_list:
        for j in gbv.plant_list:
            X_bar_j[i, j] = X_j_po_list[i, j]
            for t in range(len(gbv.period_list) + 1):
                for j2 in gbv.plant_list:
                    if t >= gbv.transit_time.get((i, j2, j), len(gbv.period_list) + 1):
                        X_bar_j[i, j][t] += X_j_po_list[i, j2][t - gbv.transit_time[i, j2, j]]

    gbv.X_outside_ub = X_po_list
    gbv.X_outside_ub_plant = X_bar_j

    return gbv


def analyze_bom(gbv):
    consistuent_list = {}   # children
    parent_list = {}    # parent
    composition_list = {}

    for i in gbv.item_list:
        consistuent_list[i] = []
        parent_list[i] = []
        for j in gbv.bom_key.keys():
            for i1, i2 in gbv.bom_key[j]:
                if (i1 == i) and (not (i2 in consistuent_list[i])):
                    consistuent_list[i].append(i2)
                if (i2 == i) and (not (i1 in parent_list[i])):
                    parent_list[i].append(i1)

    degree_list = np.zeros(len(gbv.item_list))
    keep_iter = True
    dg_no = 0.0
    while keep_iter:
        now_degree = copy.deepcopy(degree_list)
        for j in gbv.bom_key.keys():
            for i1, i2 in gbv.bom_key[j]:
                if degree_list[gbv.item_list.index(i2)] == dg_no:
                    now_degree[gbv.item_list.index(i1)] = max(now_degree[gbv.item_list.index(i1)],
                                                              degree_list[gbv.item_list.index(i2)] + 1)
                    composition_list[i1, i2] = gbv.bom_dict[j][i1, i2]
                    current_dict_keys = list(composition_list.keys())
                    for item in current_dict_keys:
                        if item[0] == i2:
                            composition_list[i1, item[1]] = gbv.bom_dict[j][i1, i2] * composition_list[item]

        if np.array_equiv(now_degree, degree_list):
            keep_iter = False
        else:
            degree_list = now_degree
            dg_no += 1

    subsidiary_list = {}
    precursor_list = {}
    for i in gbv.item_list:
        subsidiary_list[i] = [item[1] for item in composition_list.keys() if item[0] == i]
        precursor_list[i] = [item[0] for item in composition_list.keys() if item[1] == i]

    gbv.degree_list = degree_list
    gbv.consistuent_list = consistuent_list    # children list
    gbv.parent_list = parent_list          # parent list
    gbv.composition_list = composition_list    # bom production relation (quantity) between two nodes on a path in the sub-tree
    gbv.subsidiary_list = subsidiary_list    # all nodes in sub-tree
    gbv.precursor_list = precursor_list     # all precursors

    return gbv

File: src/julia/readin.jl
# script to read in the data from the csv files

using CSV
using DataFrames

readin_csv(file_add::AbstractString) = CSV.read(file_add, DataFrame)

function _id_map(values)
    return Dict(v => i - 1 for (i, v) in enumerate(values))
end

function readin_array(path_file::AbstractString)
    item_data = readin_csv(joinpath(path_file, "dm_df_item.csv"))
    unique!(item_data, :item_code)
    item_list = collect(item_data.item_code)
    I_norm = length(item_list)
    item_to_idx = _id_map(item_list)
    item_df = DataFrame(index = item_list, I = collect(0:I_norm - 1))

    plant_data = readin_csv(joinpath(path_file, "dm_df_plant.csv"))
    unique!(plant_data, :plant)

    item_plant_data = readin_csv(joinpath(path_file, "dm_df_item_plant.csv"))
    unique!(item_plant_data)

    period_data = readin_csv(joinpath(path_file, "dm_df_periods.csv"))
    unique!(period_data)
    period_list_data = period_data[period_data.length .== 1, :]
    period_list = collect(period_list_data.period)
    T_norm = length(period_list)
    period_to_idx = _id_map(period_list)
    period_df = DataFrame(index = period_list, I = collect(0:T_norm - 1))

    init_inv_data = readin_csv(joinpath(path_file, "dm_df_inv.csv"))
    unique!(init_inv_data)
    init_inv_data = init_inv_data[in.(init_inv_data.item_code, Ref(Set(item_list))), :]

    po_data = readin_csv(joinpath(path_file, "dm_df_po.csv"))
    unique!(po_data)
    po_data = po_data[in.(po_data.period, Ref(Set(period_list))) .& in.(po_data.item_code, Ref(Set(item_list))), :]

    demand_data = readin_csv(joinpath(path_file, "dm_df_demand.csv"))
    unique!(demand_data)
    demand_data = demand_data[in.(demand_data.period, Ref(Set(period_list))) .& in.(demand_data.item_code, Ref(Set(item_list))), :]

    alternate_data = readin_csv(joinpath(path_file, "dm_df_alternate_item.csv"))
    unique!(alternate_data)
    alternate_data = alternate_data[alternate_data.item_code1 .!= alternate_data.item_code2, :]
    alternate_data = alternate_data[in.(alternate_data.period, Ref(Set(period_list))), :]
    alternate_data = alternate_data[in.(alternate_data.item_code1, Ref(Set(item_list))) .& in.(alternate_data.item_code2, Ref(Set(item_list))), :]

    unit_cap_data = readin_csv(joinpath(path_file, "dm_df_unit_capacity.csv"))
    unique!(unit_cap_data)
    unit_cap_data = unit_cap_data[in.(unit_cap_data.item_code, Ref(Set(item_list))), :]
    unit_cap_dict = Dict((r.item_code, r.capacity_type) => r.unit_capacity for r in eachrow(unit_cap_data))

    set_data = readin_csv(joinpath(path_file, "dm_df_set.csv"))
    unique!(set_data)
    set_data = set_data[in.(set_data.item_code, Ref(Set(item_list))), :]
    set_list = unique(set_data.set)
    Ct_norm = length(set_list)
    set_to_idx = _id_map(set_list)
    set_df = DataFrame(index = set_list, I = collect(0:Ct_norm - 1))

    item_set_data = readin_csv(joinpath(path_file, "dm_df_item_set.csv"))
    unique!(item_set_data)
    item_set_data = item_set_data[in.(item_set_data.item_code, Ref(Set(item_list))), :]

    production_data = readin_csv(joinpath(path_file, "dm_df_production.csv"))
    unique!(production_data)
    production_data = production_data[in.(production_data.item_code, Ref(Set(item_list))), :]

    bom_data = readin_csv(joinpath(path_file, "dm_df_bom.csv"))
    unique!(bom_data)
    bom_data = bom_data[in.(bom_data.assembly, Ref(Set(item_list))) .& in.(bom_data.component, Ref(Set(item_list))), :]

    prod_plant_set = Set(production_data.plant)
    alt_plant_set = Set(alternate_data.plant)
    po_plant_set = Set(po_data.plant)
    init_inv_plant_set = Set(init_inv_data.plant)
    used_plants = unique(vcat(collect(prod_plant_set), collect(alt_plant_set), collect(po_plant_set), collect(init_inv_plant_set)))
    plant_data_used = plant_data[in.(plant_data.plant, Ref(Set(used_plants))), :]
    J_norm = nrow(plant_data_used)
    plant_to_idx = _id_map(collect(plant_data_used.plant))
    plant_df = DataFrame(index = collect(plant_data_used.plant), I = collect(0:J_norm - 1))

    cap_mask = in.(set_data.plant, Ref(Set(plant_df.index))) .& in.(set_data.item_code, Ref(Set(item_df.index))) .& in.(set_data.set, Ref(Set(set_df.index)))
    cap_rows = set_data[cap_mask, :]

    transit_data = readin_csv(joinpath(path_file, "dm_df_transit.csv"))
    transit_data = transit_data[in.(transit_data.item_code, Ref(Set(item_list))), :]
    unique!(transit_data, [:item_code, :src_plant, :dest_plant])
    transit_data = transit_data[in.(transit_data.src_plant, Ref(Set(plant_data_used.plant))) .& in.(transit_data.dest_plant, Ref(Set(plant_data_used.plant))), :]
    transit_data.transit_cost = coalesce.(transit_data.transit_cost, 0.0)
    transit_data.holding_cost = coalesce.(transit_data.holding_cost, 0.0)
    Tr_norm = nrow(transit_data)

    plant_cap_data = readin_csv(joinpath(path_file, "dm_df_max_capacity.csv"))
    unique!(plant_cap_data)
    plant_cap_data = plant_cap_data[in.(plant_cap_data.period, Ref(Set(period_list))) .&
                                    in.(plant_cap_data.plant, Ref(Set(plant_data_used.plant))) .&
                                    in.(plant_cap_data.set, Ref(Set(set_list))), :]

    df_cost = DataFrame(
        I = [item_to_idx[c] for c in item_data.item_code],
        HC = item_data.holding_cost,
        PC = item_data.holding_cost .* 1000,
    )

    alt_I_codes = Vector{eltype(alternate_data.item_code1)}(undef, nrow(alternate_data))
    alt_II_codes = Vector{eltype(alternate_data.item_code2)}(undef, nrow(alternate_data))
    for idx in 1:nrow(alternate_data)
        if alternate_data.priority_col[idx] == 2
            alt_I_codes[idx] = alternate_data.item_code1[idx]
            alt_II_codes[idx] = alternate_data.item_code2[idx]
        else
            # Match Python construction:
            # priority 1 rows are reversed (I <- item_code2, II <- item_code1).
            alt_I_codes[idx] = alternate_data.item_code2[idx]
            alt_II_codes[idx] = alternate_data.item_code1[idx]
        end
    end

    df_alt = DataFrame(
        I = [item_to_idx[c] for c in alt_I_codes],
        II = [item_to_idx[c] for c in alt_II_codes],
        J = [plant_to_idx[p] for p in alternate_data.plant],
        Ti = [period_to_idx[t] for t in alternate_data.period],
        V = alternate_data.multiplier,
    )
    unique!(df_alt)

    df_iniInv = DataFrame(
        I = [item_to_idx[c] for c in init_inv_data.item_code],
        J = [plant_to_idx[p] for p in init_inv_data.plant],
        V = init_inv_data.qty,
    )

    df_external = DataFrame(
        I = [item_to_idx[c] for c in po_data.item_code],
        J = [plant_to_idx[p] for p in po_data.plant],
        Ti = [period_to_idx[t] for t in po_data.period],
        V = po_data.qty,
    )

    df_unitCap = DataFrame(
        I = [item_to_idx[c] for c in cap_rows.item_code],
        J = [plant_to_idx[p] for p in cap_rows.plant],
        Ct = [set_to_idx[s] for s in cap_rows.set],
        V = [unit_cap_dict[(r.item_code, r.capacity_type)] for r in eachrow(cap_rows)],
    )

    df_prod = DataFrame(
        I = [item_to_idx[c] for c in production_data.item_code],
        J = [plant_to_idx[p] for p in production_data.plant],
        LT = production_data.lead_time,
        pKey = ones(Int, nrow(production_data)),
        MaxProd = production_data.max_production,
        LS = production_data.lot_size,
    )

    df_transit = DataFrame(
        I = [item_to_idx[c] for c in transit_data.item_code],
        J = [plant_to_idx[p] for p in transit_data.src_plant],
        JJ = [plant_to_idx[p] for p in transit_data.dest_plant],
        Tr = collect(0:Tr_norm - 1),
        V = ones(Int, Tr_norm),
    )

    df_transitTC = DataFrame(
        Tr = collect(0:Tr_norm - 1),
        V = transit_data.lead_time,
        TC = transit_data.transit_cost,
    )

    df_bom = DataFrame(
        II = [item_to_idx[c] for c in bom_data.component],
        I = [item_to_idx[c] for c in bom_data.assembly],
        J = [plant_to_idx[p] for p in bom_data.plant],
        V = bom_data.qty,
    )

    df_maxCap = DataFrame(
        J = [plant_to_idx[p] for p in plant_cap_data.plant],
        Ti = [period_to_idx[t] for t in plant_cap_data.period],
        Ct = [set_to_idx[s] for s in plant_cap_data.set],
        V = plant_cap_data.max_capacity,
    )

    df_demand = DataFrame(
        I = [item_to_idx[c] for c in demand_data.item_code],
        Ti = [period_to_idx[t] for t in demand_data.period],
        V = demand_data.order_demand,
    )

    return (
        df_cost = df_cost,
        df_alt = df_alt,
        df_iniInv = df_iniInv,
        df_external = df_external,
        df_unitCap = df_unitCap,
        df_prod = df_prod,
        df_transit = df_transit,
        df_transitTC = df_transitTC,
        df_bom = df_bom,
        df_maxCap = df_maxCap,
        df_demand = df_demand,
        item_df = item_df,
        plant_df = plant_df,
        period_df = period_df,
        set_df = set_df,
        I_norm = I_norm,
        J_norm = J_norm,
        T_norm = T_norm,
        Ct_norm = Ct_norm,
        Tr_norm = Tr_norm,
        item_data = item_data,
    )
end